### PR TITLE
JULY 2018 Security Update 

### PR DESCRIPTION
**Pull request by atulkatti**
JULY 2018 Security Update that addresses the following issues in ChakraCore.dll:
[CVE-2018-8275], [CVE-2018-8276], [CVE-2018-8279], [CVE-2018-8280], [CVE-2018-8283], [CVE-2018-8286], [CVE-2018-8287], [CVE-2018-8288], [CVE-2018-8290], [CVE-2018-8291], [CVE-2018-8294], [CVE-2018-8298]

### CODE REVIEW

**Comment by akroshg** [comment]
LGTM

File: lib/Runtime/Types/DictionaryPropertyDescriptor.h [hunk 1]
```diff
@@ -274,8 +274,13 @@ namespace Js
         this->Setter = (descriptor.Setter == DictionaryPropertyDescriptor<TPropertyIndexFrom>::NoSlots) ? NoSlots : descriptor.Setter;
         this->IsAccessor = descriptor.IsAccessor;
 
+        // Not strictly required, PreventFalseReference must always be 1
+        this->PreventFalseReference = descriptor.PreventFalseReference;
+        this->IsShadowed = descriptor.IsShadowed;
+
 #if ENABLE_FIXED_FIELDS
         this->IsInitialized = descriptor.IsInitialized;
+        this->IsOnlyOneAccessorInitialized = descriptor.IsOnlyOneAccessorInitialized;
```
Comment (atulkatti) [inline lib/Runtime/Types/DictionaryPropertyDescriptor.h]: [CVE-2018-8291] LGTM

File: lib/Runtime/Types/PathTypeHandler.cpp [hunk 2]
```diff
@@ -1452,7 +1452,11 @@ namespace Js
                 DictionaryPropertyDescriptor<PropertyIndex> *descriptor;
                 bool result = newTypeHandler->propertyMap->TryGetReference(propertyRecord, &descriptor);
                 Assert(result);
-                Assert(attributes[descriptor->GetDataPropertyIndex<false>()] & ObjectSlotAttr_Accessor);
+                if (!(attributes[descriptor->GetDataPropertyIndex<false>()] & ObjectSlotAttr_Accessor))
+                {
+                    // Setter without a getter; this is a stale entry, so ignore it
+                    continue;
+                }
```
Comment (atulkatti) [inline lib/Runtime/Types/PathTypeHandler.cpp]: [CVE-2018-8283]: LGTM

File: lib/Runtime/Library/DataView.cpp [hunk 3]
```diff
@@ -116,6 +116,12 @@ namespace Js
             mappedLength = byteLength - offset;
         }
 
+        // Evaluation of the argument(s) above is reentrant and can detach the array.
+        if (arrayBuffer->IsDetached())
+        {
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_DetachedTypedArray);
+        }
+
         //10.   Let O be OrdinaryCreateFromConstructor(NewTarget, "%DataViewPrototype%", [[DataView]], [[ViewedArrayBuffer]], [[ByteLength]], [[ByteOffset]]).
```
Comment (atulkatti) [inline lib/Runtime/Library/DataView.cpp]: [CVE-2018-8280] LGTM

File: lib/Runtime/Base/ThreadServiceWrapperBase.cpp [hunk 4]
```diff
@@ -78,6 +78,13 @@ bool ThreadServiceWrapperBase::ScheduleIdleCollect(uint ticks, bool scheduleAsTa
 
 bool ThreadServiceWrapperBase::IdleCollect()
 {
+    // Tracking service does not AddRef/Release the thread service and only keeps a function pointer and context parameter (this pointer)
+    // to execute the IdleCollect callback. It is possible that the tracking service gets destroyed as part of the collection
+    // during this IdleCollect. If that happens then we need to make sure ThreadService (which may be owned by the tracking service)
+    // is kept alive until this callback completes. Any pending timer is killed in the thread service destructor so we should not get
+    // any new callbacks after the thread service is destroyed.
+    AutoAddRefReleaseThreadService autoThreadServiceKeepAlive(this);
+
```
Comment (atulkatti) [inline lib/Runtime/Base/ThreadServiceWrapperBase.cpp]: [CVE-2018-8287] LGTM

File: lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp [hunk 5]
```diff
@@ -1543,7 +1536,7 @@ DEFINE_ISXLOCALEAVAILABLE(PR, uloc)
         }
 
         state->SetInternalProperty(
-            InternalPropertyIds::HiddenObject,
+            InternalPropertyIds::CachedUNumberFormat,
```
Comment (atulkatti) [inline lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: >CachedUNumberFormat [](start = 33, length = 19)

[CVE-2018-8298]  LGTM

File: lib/Backend/JnHelperMethod.cpp [hunk 6]
```diff
@@ -283,6 +283,7 @@ DECLSPEC_GUARDIGNORE  _NOINLINE intptr_t GetNonTableMethodAddress(ThreadContextI
 ///----------------------------------------------------------------------------
 intptr_t GetMethodOriginalAddress(ThreadContextInfo * context, JnHelperMethod helperMethod)
 {
+    AssertOrFailFast(helperMethod >= 0 && helperMethod < IR::JnHelperMethodCount);
```
Comment (atulkatti) [inline lib/Backend/JnHelperMethod.cpp]: >AssertOrFailFast [](start = 4, length = 16)

[CVE-2018-8276] LGTM

**Pull request merged**